cabal-ghc92.project

```diff
@@ -39,18 +39,26 @@ package *
 
 write-ghc-environment-files: never
 
-index-state: 2022-04-19T07:18:40Z
+index-state: 2022-04-25T10:14:27Z
 
 constraints:
   -- These plugins don't build/work on GHC92 yet
   haskell-language-server
     +ignore-plugins-ghc-bounds
     -brittany
     -haddockComments
-    -hlint
     -retrie
     -splice
     -tactic,
+  hls-hlint-plugin +ghc-lib,
+  ghc-lib-parser-ex -auto,
+  hlint +ghc-lib,
+  stylish-haskell +ghc-lib
+
+source-repository-package
```

Comment:
Why do we need this? when can we remove it?

Reply:
Due to the limitation of `hls-hlint-plugin` code structure, we must force `ghc-lib-parser` uses `ghc-lib` instead of ghc to make compile, this repo adds a flag to allow us to force use `ghc-lib` in `stylish-haskell`(or is there another way to do this with out edit the package?)

---
> Why do we have to use `ghc-lib`?

1. We have two code path in hls-hlint-plugin, includes using `ghc` and using `ghc-lib`. the `ghc` codebase relies on our compiled module by `getParsedModuleWithComments` which has some `ann` operation that ghc-9.2 has totally different views, we can make `ghc-9.2` also use this after some revise(Not much work, but will introduce more complex compat).  The `ghc-lib` codebase parses the source code directly and we can use this for any ghc-lib easily.
2. We use `apply-refact` to apply hlint, and for ghc before 9.2, it requires `apply-refact-0.9` and in ghc-9.2, it needs `apply-refact-0.10`. Since in `ghc` codebase we pass parsed module to apply-refact, it works in 0.9 but not in 0.10.  `apply-refact-0.10` uses an internal `IORef`(https://github.com/mpickering/apply-refact/blob/cb8464af74ce9c1cee26fa4862c713bb5ff96114/src/Refact/Internal.hs#L751-L755) and the only entrance is parse the source by `apply-refact`. (https://github.com/mpickering/apply-refact/blob/cb8464af74ce9c1cee26fa4862c713bb5ff96114/src/Refact/Internal.hs#L698)

---
I think we can remove this after the new `stylish-haskell` with `ghc-lib` flag is released.

Reply:
It looks we can turn to `ghc-lib` and abandon `ghc` in `hls-hlint-plugin`, can we?

Reply:
Pr created https://github.com/haskell/stylish-haskell/pull/402